README.md
## Coupled parameterized reduced order modelling of thermo-mechanical phenomena arising in blast furnace ##

### 0. Introduction

This repository contains numerical implementation of the benchmark tests reported in "Coupled parameterized reduced order modelling of thermo-mechanical phenomena arising in blast furnace" (see https://zenodo.org/record/3888145). 

### 1. Prerequisites

We use **python** 3.6.9 as the programming language. In this project we use the libraries :
* **FEniCS** 2019.1.0 (www.fenicsproject.org)
* **RBniCS** 0.1.dev1 (www.rbnicsproject.org)
* **Matplolib** 3.1.2 (www.matplotlib.org)
* **numpy** 1.17.4 (www.numpy.org)

The solutions are stored in **.pvd** format, which can later be viewed with **Paraview** (www.paraview.org).

### 2. Installation

Simply clone the public repository:

```
git clone https://github.com/ROMSOC/benchmark_thermomechanical_model
```

### 3. Running the benchmark cases

Source codes for input data are provided in the folder "source_files". Source codes for running the benchmark are provided in folder "source". After running the benchmark, results are stored in folder "result_files".

Run required .py file e.g. "file_name.py" as,
```
python3 file_name.py
```

### 4. Authors and contributors

This code has been developed by [Nirav Vasant Shah] [email](mailto:shah.nirav@sissa.it) under the supervision of [Dr. Michele Girfoglio] [email](mailto:michele.girfoglio@sissa.it), [Dr. Patricia Barral] [email](mailto:patricia.barral@usc.es), [Prof. Peregrina Quintela] [email](mailto:peregrina.quintela@itmati.com), [Prof. Gianluigi Rozza] [email](mailto:gianluigi.rozza@sissa.it) and [Ing. Alejandro Lengomin] [email](mailto:alejandro.lengomin@arcelormittal.com).

### 5. How to cite

	@misc{thermomechanical_coupled_one_way-0.1,
	key          = {ThermoMechanicalCoupled},
	author       = {Shah, N.V. and Girfoglio, M. and Barral, P. and Quintela, P. and Rozza, G. and Lengomin, A.},
	title        = {{ThermoMechanicalCoupled} -- {Coupled parameterized reduced order modelling of thermo-mechanical phenomena arising in blast furnaces} {Version 0.1}},
	month        = June,
	year         = 2021,
	doi          = {10.xxxx/zenodo.xxxxxxx},
	note         = {see also: \url{https://zenodo.org/record/3888145}}
	}

### 6. License

* **FEniCS** and **RBniCS** are freely available under the GNU LGPL, version 3.
* **Matplotlib** only uses BSD compatible code, and its license is based on the PSF license. Non-BSD compatible licenses (e.g., LGPL) are acceptable in matplotlib toolkits.

Accordingly, this code is freely available under the GNU LGPL, version 3.

### 7. Disclaimer
In downloading this SOFTWARE you are deemed to have read and agreed to the following terms: This SOFT- WARE has been designed with an exclusive focus on civil applications. It is not to be used for any illegal, deceptive, misleading or unethical purpose or in any military applications. This includes ANY APPLICATION WHERE THE USE OF THE SOFTWARE MAY RESULT IN DEATH, PERSONAL INJURY OR SEVERE PHYSICAL OR ENVIRONMENTAL DAMAGE. Any redistribution of the software must retain this disclaimer. BY INSTALLING, COPYING, OR OTHERWISE USING THE SOFTWARE, YOU AGREE TO THE TERMS ABOVE. IF YOU DO NOT AGREE TO THESE TERMS, DO NOT INSTALL OR USE THE SOFTWARE.


CHANGELOG.md
# June 30, 2021 
* First release


source/thermal_model/benchmark_thermal.py
from __future__ import print_function
from dolfin import *
import matplotlib.pyplot as plt

# Read the mesh file from specified path.
mesh = Mesh("../../benchmarks/data_files/mesh_data/hearth.xml")
domains = MeshFunction('size_t',mesh,mesh.topology().dim()) # Read domain marker
subdomains = MeshFunction("size_t", mesh, "../../benchmarks/data_files/mesh_data/hearth_physical_region.xml") # Read subdomain markers
boundaries = MeshFunction("size_t", mesh, "../../benchmarks/data_files/mesh_data/hearth_facet_region.xml") # Read boundary markers
dx = Measure('dx', domain = mesh, subdomain_data = domains) # Volume measure
ds = Measure('ds', domain = mesh, subdomain_data = boundaries) # Boundary measure
n = as_vector(FacetNormal(mesh)) # Edge unit normal vector

d_bottom = ds(2) + ds(3) + ds(4) + ds(5) + ds(6) # Markers of bottom boundary \gamma_{-}
d_out = ds(7) + ds(8) + ds(9) + ds(10) + ds(11) # Markers of outer boundary \gamma_{out}
d_sf = ds(13) + ds(14) + ds(15) + ds(16) + ds(17) + ds(18) + ds(19) + ds(20) # Markers of inner boundary \gamma_{sf}

#Computation of H^1_r(\omega) norm
def compute_h1r_norm(psi,mesh):
	r = SpatialCoordinate(mesh)[0]
	dx = Measure('dx', domain = mesh)
	a = inner(psi,psi)*r*dx + inner(grad(psi),grad(psi))*r*dx
	A = assemble(a)
	return sqrt(A)

error_T_vector = [] #List to store error in temperature w.r.t. polynomial degree
p = range(1,4) # List of polynomial degrees

# Known analytical solution, Thermal material properties and Boundary data
T_analytical = Expression('x[0]*x[0]*x[1]',degree = 3) 
VT_analytical = FunctionSpace(mesh,"CG",3) #Space for analytical solution
T_analytical = project(T_analytical,VT_analytical)
k = 10. # Thermal conductivity
h_fluid = 200. # Convection coefficient on \gamma_{sf}
h_right = 2000. # Convection coefficient on \gamma_{out}
h_bottom = 2000. # Convection coefficient on \gamma_{-}

for i in p:
	# Define function space
	VT = FunctionSpace(mesh,"CG",i) # Function space for temperature
	psi, T_ = TestFunction(VT), TrialFunction(VT) # Evaluate trial and test function
	T = Function(VT, name = "temperature increase")
	x = list()
	x.append(Expression("x[0]", element=VT.ufl_element())) #r coordinate
	x.append(Expression("x[1]", element=VT.ufl_element())) #y coordinate

	# solving weak form of energy equation
	a_T = k * inner(grad(psi),grad(T_)) * x[0] * dx + \
		h_fluid * psi * T_ * x[0] * d_sf + h_right * psi * T_ * x[0] * d_out + \
		h_bottom * psi * T_ * x[0] * d_bottom # Bilinear side
	l_T = h_fluid * psi * (x[0] * x[0] * x[1] + k/h_fluid * ( 2 * x[0] * x[1] * n[0] + x[0] * x[0] * n[1] ) ) * x[0] * d_sf + \
		h_right * psi * (x[0]*x[0]*x[1]+2*x[0]*x[1]*k/h_right) * x[0] * d_out + \
		h_bottom * psi * (x[0] * x[0] * x[1] - x[0] * x[0] * k / h_bottom) * x[0] * d_bottom + \
		-4 * k * x[1] * psi * x[0] * dx + psi * k * x[0] * x[0] * x[0] * ds(12) # Linear side
	solve(a_T == l_T, T) # Solve the variational form

	#Computing temperature relative error in relevant norm
	error_T = compute_h1r_norm(T_analytical-T,mesh)/compute_h1r_norm(T,mesh)
	error_T_vector.append(error_T)
	print("Temperature relative error in H^1_r norm : " + str(error_T))

# Plotting and visualization
File("../../benchmarks/result_files/thermal_model/temperature_computed.pvd") << T
File("../../benchmarks/result_files/thermal_model/temperature_analytical.pvd") << T_analytical
error_temperature = Function(VT) #Function for Spatial distribution of temperature absolute error
error_temperature.vector()[:] = abs(T_analytical.vector().get_local() - T.vector().get_local())
File("../../benchmarks/result_files/thermal_model/temperature_absolute_error.pvd") << error_temperature

# Plotting and printing convergence tests
plt.figure(figsize=[10,8])
a = plt.semilogy([1,2,3],error_T_vector,marker='o',linewidth=4)
plt.xticks([1,2,3],fontsize=18)
plt.yticks(fontsize=18)
plt.xlabel('Polynomial degree',fontsize=24)
plt.ylabel('Relative error',fontsize=24)
plt.axis('tight')
plt.savefig("../../benchmarks/result_files/thermal_model/convergence_test")
plt.show()
print("Relative error in H^1_r norm: "+ str(error_T_vector))


source/mechanical_model/benchmark_mechanical.py
from __future__ import print_function
from dolfin import *
import matplotlib.pyplot as plt
import numpy as np

# Read the mesh file from specified path.
mesh = Mesh("../../benchmarks/data_files/mesh_data/hearth.xml")
domains = MeshFunction('size_t',mesh,mesh.topology().dim()) # Read domain marker
subdomains = MeshFunction("size_t", mesh, "../../benchmarks/data_files/mesh_data/hearth_physical_region.xml") # Read subdomain markers
boundaries = MeshFunction("size_t", mesh, "../../benchmarks/data_files/mesh_data/hearth_facet_region.xml") # Read boundary markers
dx = Measure('dx', domain = mesh, subdomain_data = domains) # Volume measure
ds = Measure('ds', domain = mesh, subdomain_data = boundaries) # Boundary measure
n = as_vector(FacetNormal(mesh)) # Edge unit normal vector

d_bottom = ds(2) + ds(3) + ds(4) + ds(5) + ds(6) # Markers of bottom boundary \gamma_{-}
d_out = ds(7) + ds(8) + ds(9) + ds(10) + ds(11) # Markers of outer boundary \gamma_{out}
d_sf = ds(13) + ds(14) + ds(15) + ds(16) + ds(17) + ds(18) + ds(19) + ds(20) # Markers of inner boundary \gamma_{sf}

#Computation of \mathbb{U} norm
def compute_U_norm(phi,mesh):
	x = SpatialCoordinate(mesh)
	dx = Measure('dx', domain = mesh)
	a = inner(phi,phi)*x[0]*dx + inner(grad(phi),grad(phi))*x[0]*dx + (phi[0]**2)/x[0]*dx
	A = assemble(a)
	return sqrt(A)

# Axisymmetric strain tensor definition. Alternative could be to express strain as vector using Voigt notation.
def eps(u): 
	return \
		sym(as_tensor([[u[0].dx(0), u[0].dx(1), 0. ],\
		[u[1].dx(0), u[1].dx(1), 0.],\
		[0., 0., u[0]/x[0]]]))

# Axisymmetric stress tensor definition. Alternative could be to express stress as vector using Voigt notation.
def sigma(u):
	return lmbda * tr(eps(u)) * Identity(3) + 2.0 * mu * eps(u)

error_u_vector = [] #List for absolute error in displacement
p = range(1,4) #Polynomial degrees

# Known analytical solution, Mechanical material properties and Boundary data
u_analytical = Expression(("A * x[0] * x[1] * x[1]","A * x[0] * x[0] * x[1]"),A=1e-4,degree=3) 
# analytical solution started as expression and projected onto relevant funtion space
VM_analytical = VectorFunctionSpace(mesh,"CG",3)
u_analytical = project(u_analytical,VM_analytical)
E = Constant(5e9) # Young's modulus
nu = Constant(0.2) # Poisson's ratio
mu = E/2/(1+nu) # Lam\'e parameter
lmbda = E*nu/(1+nu)/(1-2*nu) # Lam\'e parameter

for i in p:
	# Define function space
	VM = VectorFunctionSpace(mesh,"CG",i) # Function space for displacement
	x = Expression(("x[0]","x[1]"), element=VM.ufl_element())
	VS = FunctionSpace(mesh,"CG",max(i-1,1)) 
	# Function space for Von Mises stress NOTE: when i=1, the VS is of degree 1 and not 0.
	phi, u_ = TestFunction(VM), TrialFunction(VM)
	u = Function(VM, name = "Displacement") # u[0] = u_r and u[1] = u_y

	# Dirichlet boundary data
	bcs_M = [DirichletBC( VM.sub(0), Constant(0.), 'x[0] < DOLFIN_EPS and on_boundary'), \
		DirichletBC( VM.sub(1), Constant(0.), 'near(x[1],0) and on_boundary')] 
	# Set Dirichlet boundary. Note that only functions which satisfy zero normal displacement on \gamma_s \cup \gamma_- are admissible.

	# Source term and relevant boundary data
	f0_r = - (2*E*nu*1e-4*x[0]/(1-2*nu)/(1+nu)+2*E*1e-4*x[0]/(1+nu)) 
	f0_y = - (4*E*1e-4*x[1]/(1+nu)+4*E*1e-4*x[1]*nu/(1-2*nu)/(1+nu)) 

	g_plus_r = 2*E*1e-4*x[0]*x[1]/(1+nu)
	g_plus_y = E / (1-2*nu) / (1+nu) * (2*nu*1e-4*x[1]*x[1]+(1-nu)*1e-4*x[0]*x[0])

	g_minus_r = -g_plus_r

	g_sf_r = E / (1-2*nu) / (1+nu) * (1e-4 * x[1] * x[1] + nu * 1e-4 * x[0] * x[0]) * n[0] + 2 * E * 1e-4 * x[0] * x[1] / (1 + nu) * n[1]
	g_sf_y = 2 * E * 1e-4 * x[0] * x[1] / (1 + nu) * n[0] + E / (1-2*nu) / (1+nu) * (2 * nu * 1e-4 * x[1] * x[1] + (1 - nu) * 1e-4 * x[0] * x[0]) * n[1]

	g_out_r = E / (1-2*nu) / (1+nu) * (1e-4 * x[1] * x[1] + nu * 1e-4 * x[0] * x[0])
	g_out_y = 2*E*1e-4*x[0]*x[1]/(1+nu)

	# solving weak form of momentum equation
	# Bilinear form
	a_M = inner(sigma(u_),eps(phi)) * x[0] * dx
	#linear form
	l_M = (phi[0] * f0_r + phi[1] * f0_y) * x[0] * dx + (phi[0] * g_plus_r + phi[1] * g_plus_y) * x[0] * ds(12) + \
	(phi[0] * g_minus_r) * x[0] * d_bottom + (phi[0] * g_sf_r + phi[1] * g_sf_y) * x[0] * d_sf + \
	(phi[0] * g_out_r + phi[1] * g_out_y) * x[0] * d_out
	solve(a_M == l_M, u, bcs_M)

	# Von Mises stress computed displacement
	sigma_dev = sigma(u) - tr(sigma(u)) / 3 * Identity(3) 
	sigma_vm = sqrt(3 * inner( sigma_dev, sigma_dev) / 2) # Von mises stress
	# Von Mises stress analytical displacement
	sigma_dev_analytical = sigma(u_analytical) - tr(sigma(u_analytical)) / 3 * Identity(3)
	sigma_vm_analytical = sqrt(3 * inner( sigma_dev_analytical, sigma_dev_analytical) / 2) # Von mises stress

	# Compute H^1_r norm of error
	error_u = compute_U_norm(u_analytical-u,mesh)/compute_U_norm(u_analytical,mesh)
	error_u_vector.append(error_u)
	print("Relative error in U-norm : ",str(error_u))

# Post-processing and visualization
File("../../benchmarks/result_files/mechanical_model/displacement_computed.pvd") << u
File("../../benchmarks/result_files/mechanical_model/displacement_analytical.pvd") << u_analytical
File("../../benchmarks/result_files/mechanical_model/displacement_error.pvd") << project(u_analytical-u,VM)

error_stress = Function(VS) #Function for absolute error in stress tensor
error_stress.vector()[:] = abs(project(sigma_vm,VS).vector().get_local() - project(sigma_vm_analytical,VS).vector().get_local())
File("../../benchmarks/result_files/mechanical_model/von_mises_stress_computed.pvd") << project(sigma_vm,VS)
File("../../benchmarks/result_files/mechanical_model/von_mises_stress_analytical.pvd") << project(sigma_vm_analytical,VS)
File("../../benchmarks/result_files/mechanical_model/von_mises_stress_error.pvd") << project(error_stress,VS)

#Convergence tests
plt.figure(figsize=[10,8])
a = plt.semilogy([1,2,3],error_u_vector,marker='o',linewidth=4)
plt.xticks([1,2,3],fontsize=18)
plt.yticks(fontsize=18)
plt.xlabel('Polynomial degree',fontsize=24)
plt.ylabel('Relative error',fontsize=24)
plt.axis('tight')
plt.savefig("../../benchmarks/result_files/mechanical_model/convergence_test")
plt.show() # To show the plots

print("Relative error in U norm: "+ str(error_u_vector))


benchmarks/source_files/generate_mesh_hearth.py
# Copyright (C) 2015-2019 by the RBniCS authors
#
# This file is part of RBniCS.
#
# RBniCS is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# RBniCS is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with RBniCS. If not, see <http://www.gnu.org/licenses/>.
#

from dolfin import * #FEniCS library
from mshr import * #mshr - mesh generation component of FEniCS
from rbnics import * #RBniCS library
from rbnics.backends.dolfin.wrapping import counterclockwise
from rbnics.shape_parametrization.utils.symbolic import VerticesMappingIO
import os
import matplotlib.pyplot as plt #matplotlib library

# Define domain
domain = Polygon([Point(0.,0.),Point(7.05,0.),
				Point(7.05,7.265),Point(5.3,7.265),
				Point(5.3,4.065),Point(4.95,4.065),
				Point(4.95,3.565),Point(4.6,3.565),
				Point(4.6,2.965),Point(4.25,2.965),
				Point(4.25,2.365),Point(0.,2.365)])

# Define vertices mappings of affine shape parametrization. These will be used
# to partition the mesh in subdomains.
vertices_mappings = [
	{
		("0", "0"): ("0", "0"),
		("4.25", "0"): ("mu[6]/2", "0"),
		("0", "2.365"): ("0", "mu[0]")
	}, # subdomain 1
	{
		("0", "2.365"): ("0", "mu[0]"),
		("4.25", "0"): ("mu[6]/2", "0"),
		("4.25", "2.365"): ("mu[6]/2", "mu[0]")
	}, # subdomain 2
	{
		("4.25", "0"): ("mu[6]/2", "0"),
		("4.6", "0"): ("mu[7]/2", "0"),
		("4.25", "2.365"): ("mu[6]/2", "mu[0]")
	}, # subdomain 3
	{
		("4.25", "2.365"): ("mu[6]/2", "mu[0]"),
		("4.6", "0"): ("mu[7]/2", "0"),
		("4.6", "2.365"): ("mu[7]/2", "mu[0]")
	}, # subdomain 4
	{
		("4.6", "0"): ("mu[7]/2", "0"),
		("4.95", "0"): ("mu[8]/2", "0"),
		("4.6", "2.365"): ("mu[7]/2", "mu[0]")
	}, # subdomain 5
	{
		("4.6", "2.365"): ("mu[7]/2", "mu[0]"),
		("4.95", "0"): ("mu[8]/2", "0"),
		("4.95", "2.365"): ("mu[8]/2", "mu[0]")
	}, # subdomain 6
	{
		("4.95", "0"): ("mu[8]/2", "0"),
		("5.3", "0"): ("mu[9]/2", "0"),
		("4.95", "2.365"): ("mu[8]/2", "mu[0]")
	}, # subdomain 7
	{
		("4.95", "2.365"): ("mu[8]/2", "mu[0]"),
		("5.3", "0"): ("mu[9]/2", "0"),
		("5.3", "2.365"): ("mu[9]/2", "mu[0]")
	}, # subdomain 8
	{
		("5.3", "0"): ("mu[9]/2", "0"),
		("7.05", "0"): ("mu[5]/2", "0"),
		("5.3", "2.365"): ("mu[9]/2", "mu[0]")
	}, # subdomain 9
	{
		("5.3", "2.365"): ("mu[9]/2", "mu[0]"),
		("7.05", "0"): ("mu[5]/2", "0"),
		("7.05", "2.365"): ("mu[5]/2", "mu[0]")
	}, # subdomain 10
	{
		("4.25", "2.365"): ("mu[6]/2", "mu[0]"),
		("4.6", "2.365"): ("mu[7]/2", "mu[0]"),
		("4.25", "2.965"): ("mu[6]/2", "mu[0]+mu[1]")
	}, # subdomain 11
	{
		("4.25", "2.965"): ("mu[6]/2", "mu[0]+mu[1]"),
		("4.6", "2.365"): ("mu[7]/2", "mu[0]"),
		("4.6", "2.965"): ("mu[7]/2", "mu[0]+mu[1]")
	}, # subdomain 12
	{
		("4.6", "2.365"): ("mu[7]/2", "mu[0]"),
		("4.95", "2.365"): ("mu[8]/2", "mu[0]"),
		("4.6", "2.965"): ("mu[7]/2", "mu[0]+mu[1]")
	}, # subdomain 13
	{
		("4.6", "2.965"): ("mu[7]/2", "mu[0]+mu[1]"),
		("4.95", "2.365"): ("mu[8]/2", "mu[0]"),
		("4.95", "2.965"): ("mu[8]/2", "mu[0]+mu[1]")
	}, # subdomain 14
	{
		("4.95", "2.365"): ("mu[8]/2", "mu[0]"),
		("5.3", "2.365"): ("mu[9]/2", "mu[0]"),
		("4.95", "2.965"): ("mu[8]/2", "mu[0]+mu[1]")
	}, # subdomain 15
	{
		("4.95", "2.965"): ("mu[8]/2", "mu[0]+mu[1]"),
		("5.3", "2.365"): ("mu[9]/2", "mu[0]"),
		("5.3", "2.965"): ("mu[9]/2", "mu[0]+mu[1]")
	}, # subdomain 16
	{
		("5.3", "2.365"): ("mu[9]/2", "mu[0]"),
		("7.05", "2.365"): ("mu[5]/2", "mu[0]"),
		("5.3", "2.965"): ("mu[9]/2", "mu[0]+mu[1]")
	}, # subdomain 17
	{
		("5.3", "2.965"): ("mu[9]/2", "mu[0]+mu[1]"),
		("7.05", "2.365"): ("mu[5]/2", "mu[0]"),
		("7.05", "2.965"): ("mu[5]/2", "mu[0]+mu[1]")
	}, # subdomain 18
	{
		("4.6", "2.965"): ("mu[7]/2", "mu[0]+mu[1]"),
		("4.95", "2.965"): ("mu[8]/2", "mu[0]+mu[1]"),
		("4.6", "3.565"): ("mu[7]/2", "mu[0]+mu[1]+mu[2]")
	}, # subdomain 19
	{
		("4.6", "3.565"): ("mu[7]/2", "mu[0]+mu[1]+mu[2]"),
		("4.95", "2.965"): ("mu[8]/2", "mu[0]+mu[1]"),
		("4.95", "3.565"): ("mu[8]/2", "mu[0]+mu[1]+mu[2]")
	}, # subdomain 20
	{
		("4.95", "2.965"): ("mu[8]/2", "mu[0]+mu[1]"),
		("5.3", "2.965"): ("mu[9]/2", "mu[0]+mu[1]"),
		("4.95", "3.565"): ("mu[8]/2", "mu[0]+mu[1]+mu[2]")
	}, # subdomain 21
	{
		("4.95", "3.565"): ("mu[8]/2", "mu[0]+mu[1]+mu[2]"),
		("5.3", "2.965"): ("mu[9]/2", "mu[0]+mu[1]"),
		("5.3", "3.565"): ("mu[9]/2", "mu[0]+mu[1]+mu[2]")
	}, # subdomain 22
	{
		("5.3", "2.965"): ("mu[9]/2", "mu[0]+mu[1]"),
		("7.05", "2.965"): ("mu[5]/2", "mu[0]+mu[1]"),
		("5.3", "3.565"): ("mu[9]/2", "mu[0]+mu[1]+mu[2]")
	}, # subdomain 23
	{
		("5.3", "3.565"): ("mu[9]/2", "mu[0]+mu[1]+mu[2]"),
		("7.05", "2.965"): ("mu[5]/2", "mu[0]+mu[1]"),
		("7.05", "3.565"): ("mu[5]/2", "mu[0]+mu[1]+mu[2]")
	}, # subdomain 24
	{
		("4.95", "3.565"): ("mu[8]/2", "mu[0]+mu[1]+mu[2]"),
		("5.3", "3.565"): ("mu[9]/2", "mu[0]+mu[1]+mu[2]"),
		("4.95", "4.065"): ("mu[8]/2", "mu[0]+mu[1]+mu[2]+mu[3]")
	}, # subdomain 25
	{
		("4.95", "4.065"): ("mu[8]/2", "mu[0]+mu[1]+mu[2]+mu[3]"),
		("5.3", "3.565"): ("mu[9]/2", "mu[0]+mu[1]+mu[2]"),
		("5.3", "4.065"): ("mu[9]/2", "mu[0]+mu[1]+mu[2]+mu[3]")
	}, # subdomain 26
	{
		("5.3", "3.565"): ("mu[9]/2", "mu[0]+mu[1]+mu[2]"),
		("7.05", "3.565"): ("mu[5]/2", "mu[0]+mu[1]+mu[2]"),
		("5.3", "4.065"): ("mu[9]/2", "mu[0]+mu[1]+mu[2]+mu[3]")
	}, # subdomain 27
	{
		("5.3", "4.065"): ("mu[9]/2", "mu[0]+mu[1]+mu[2]+mu[3]"),
		("7.05", "3.565"): ("mu[5]/2", "mu[0]+mu[1]+mu[2]"),
		("7.05", "4.065"): ("mu[5]/2", "mu[0]+mu[1]+mu[2]+mu[3]")
	}, # subdomain 28
	{
		("5.3", "4.065"): ("mu[9]/2", "mu[0]+mu[1]+mu[2]+mu[3]"),
		("7.05", "4.065"): ("mu[5]/2", "mu[0]+mu[1]+mu[2]+mu[3]"),
		("5.3", "7.265"): ("mu[9]/2", "mu[0]+mu[1]+mu[2]+mu[3]+mu[4]")
	}, # subdomain 29
	{
		("5.3", "7.265"): ("mu[9]/2", "mu[0]+mu[1]+mu[2]+mu[3]+mu[4]"),
		("7.05", "4.065"): ("mu[5]/2", "mu[0]+mu[1]+mu[2]+mu[3]"),
		("7.05", "7.265"): ("mu[5]/2", "mu[0]+mu[1]+mu[2]+mu[3]+mu[4]")
	} # subdomain 30
]

# Loop over all mappings and set subdomain markers
for i, vertices_mapping in enumerate(vertices_mappings):
	print(i,vertices_mapping.keys())
	subdomain_i = Polygon([Point(*[float(coord) for coord in vertex]) for vertex in counterclockwise(vertices_mapping.keys())])
	domain.set_subdomain(i + 1, subdomain_i)

# Create mesh
mesh = generate_mesh(domain, 30) #30 specifies the mesh size.

# Create subdomains
subdomains = MeshFunction("size_t", mesh, 2, mesh.domains())

# Define classes for different boundaries
class Gamma_s(SubDomain):
	def inside(self, x, on_boundary):
		return x[0] < DOLFIN_EPS and on_boundary

class Gamma_minus1(SubDomain):
	def inside(self, x, on_boundary):
		return x[1] < DOLFIN_EPS and x[0] < (4.255 + DOLFIN_EPS) and on_boundary

class Gamma_minus2(SubDomain):
	def inside(self, x, on_boundary):
		return x[1] < DOLFIN_EPS and x[0] > (4.2 - DOLFIN_EPS) and x[0] < (4.65 + DOLFIN_EPS) and on_boundary

class Gamma_minus3(SubDomain):
	def inside(self, x, on_boundary):
		return x[1] < DOLFIN_EPS and x[0] > (4.55 - DOLFIN_EPS) and x[0] < (5.05 + DOLFIN_EPS) and on_boundary

class Gamma_minus4(SubDomain):
	def inside(self, x, on_boundary):
		return x[1] < DOLFIN_EPS and x[0] > (4.9 - DOLFIN_EPS) and x[0] < (5.4 + DOLFIN_EPS) and on_boundary

class Gamma_minus5(SubDomain):
	def inside(self, x, on_boundary):
		return x[1] < DOLFIN_EPS and x[0] > (5.25 - DOLFIN_EPS) and x[0] < (7.15 + DOLFIN_EPS) and on_boundary

class Gamma_out1(SubDomain):
	def inside(self, x, on_boundary):
		return x[0] > (7.04 - DOLFIN_EPS) and x[1] > (0. - DOLFIN_EPS) and x[1] < (2.4 + DOLFIN_EPS) and on_boundary

class Gamma_out2(SubDomain):
	def inside(self, x, on_boundary):
		return x[0] > (7.04 - DOLFIN_EPS) and x[1] > (2.3 - DOLFIN_EPS) and x[1] < (3. + DOLFIN_EPS) and on_boundary

class Gamma_out3(SubDomain):
	def inside(self, x, on_boundary):
		return x[0] > (7.04 - DOLFIN_EPS) and x[1] > (2.9 - DOLFIN_EPS) and x[1] < (3.6 + DOLFIN_EPS) and on_boundary

class Gamma_out4(SubDomain):
	def inside(self, x, on_boundary):
		return x[0] > (7.04 - DOLFIN_EPS) and x[1] > (3.5 - DOLFIN_EPS) and x[1] < (4.1 + DOLFIN_EPS) and on_boundary

class Gamma_out5(SubDomain):
	def inside(self, x, on_boundary):
		return x[0] > (7.04 - DOLFIN_EPS) and x[1] > (4. - DOLFIN_EPS) and x[1] < (7.3 + DOLFIN_EPS) and on_boundary

class Gamma_plus(SubDomain):
	def inside(self, x, on_boundary):
		return x[1] > (7.263 - DOLFIN_EPS) and x[0] > (4.94 - DOLFIN_EPS) and x[0] < (7.06 + DOLFIN_EPS) and on_boundary

class Gamma_sf1(SubDomain):
	def inside(self, x, on_boundary):
		return x[0] < 5.4 and x[0] > 5.28 and x[1] > (4.06 - DOLFIN_EPS) and x[1] < (7.27 + DOLFIN_EPS) and on_boundary

class Gamma_sf2(SubDomain):
	def inside(self, x, on_boundary):
		return x[1] < (4.07 + DOLFIN_EPS) and x[1] > (4.06 - DOLFIN_EPS) and x[0] > (4.93 - DOLFIN_EPS) and x[0] < (5.31 + DOLFIN_EPS) and on_boundary

class Gamma_sf3(SubDomain):
	def inside(self, x, on_boundary):
		return x[1] < (4.07 + DOLFIN_EPS) and x[1] > (3.56 - DOLFIN_EPS) and x[0] > (4.94 - DOLFIN_EPS) and x[0] < (4.96 + DOLFIN_EPS) and on_boundary

class Gamma_sf4(SubDomain):
	def inside(self, x, on_boundary):
		return x[0] < (4.97 + DOLFIN_EPS) and x[0] > (4.58 - DOLFIN_EPS) and x[1] > (3.55 - DOLFIN_EPS) and x[1] < (3.58 + DOLFIN_EPS) and on_boundary

class Gamma_sf5(SubDomain):
	def inside(self, x, on_boundary):
		return x[0] < (4.62 - DOLFIN_EPS) and x[0] > (4.58 + DOLFIN_EPS) and x[1] > (2.965 - DOLFIN_EPS) and x[1] < (3.565 + DOLFIN_EPS) and on_boundary

class Gamma_sf6(SubDomain):
	def inside(self, x, on_boundary):
		return x[0] < (4.62 + DOLFIN_EPS) and x[0] > (4.23 - DOLFIN_EPS) and x[1] > (2.96 - DOLFIN_EPS) and x[1] < (2.98 + DOLFIN_EPS) and on_boundary

class Gamma_sf7(SubDomain):
	def inside(self, x, on_boundary):
		return x[0] < (4.26 + DOLFIN_EPS) and x[0] > (4.24 - DOLFIN_EPS) and x[1] > (2.365 - DOLFIN_EPS) and x[1] < (2.965 + DOLFIN_EPS) and on_boundary

class Gamma_sf8(SubDomain):
	def inside(self, x, on_boundary):
		return x[0] < (4.26 + DOLFIN_EPS) and x[0] > (0. - DOLFIN_EPS) and x[1] > (2.36 - DOLFIN_EPS) and x[1] < (2.37 + DOLFIN_EPS) and on_boundary

# Set bounary markers by initializing instant of each class
boundaries = MeshFunction("size_t", mesh, mesh.topology().dim() - 1)
boundaries.set_all(0)
gamma_s = Gamma_s()
gamma_s.mark(boundaries, 1)
gamma_minus1 = Gamma_minus1()
gamma_minus1.mark(boundaries, 2)
gamma_minus2 = Gamma_minus2()
gamma_minus2.mark(boundaries, 3)
gamma_minus3 = Gamma_minus3()
gamma_minus3.mark(boundaries, 4)
gamma_minus4 = Gamma_minus4()
gamma_minus4.mark(boundaries, 5)
gamma_minus5 = Gamma_minus5()
gamma_minus5.mark(boundaries, 6)
gamma_out1 = Gamma_out1()
gamma_out1.mark(boundaries, 7)
gamma_out2 = Gamma_out2()
gamma_out2.mark(boundaries, 8)
gamma_out3 = Gamma_out3()
gamma_out3.mark(boundaries, 9)
gamma_out4 = Gamma_out4()
gamma_out4.mark(boundaries, 10)
gamma_out5 = Gamma_out5()
gamma_out5.mark(boundaries, 11)
gamma_plus = Gamma_plus()
gamma_plus.mark(boundaries, 12)
gamma_sf1 = Gamma_sf1()
gamma_sf1.mark(boundaries, 13)
gamma_sf2 = Gamma_sf2()
gamma_sf2.mark(boundaries, 14)
gamma_sf3 = Gamma_sf3()
gamma_sf3.mark(boundaries, 15)
gamma_sf4 = Gamma_sf4()
gamma_sf4.mark(boundaries, 16)
gamma_sf5 = Gamma_sf5()
gamma_sf5.mark(boundaries, 17)
gamma_sf6 = Gamma_sf6()
gamma_sf6.mark(boundaries, 18)
gamma_sf7 = Gamma_sf7()
gamma_sf7.mark(boundaries, 19)
gamma_sf8 = Gamma_sf8()
gamma_sf8.mark(boundaries, 20)

# Save mesh data
os.system("mkdir ../input_data/mesh_data")
VerticesMappingIO.save_file(vertices_mappings, ".", "../data_files/mesh_data/hearth_vertices_mapping.vmp")
File("../data_files/mesh_data/hearth.xml") << mesh
File("../data_files/mesh_data/hearth_physical_region.xml") << subdomains
File("../data_files/mesh_data/hearth_facet_region.xml") << boundaries
XDMFFile("../data_files/mesh_data/hearth.xdmf").write(mesh)
XDMFFile("../data_files/mesh_data/hearth_physical_region.xdmf").write(subdomains)
XDMFFile("../data_files/mesh_data/hearth_facet_region.xdmf").write(boundaries)
File("../data_files/mesh_data/hearth.pvd") << mesh
File("../data_files/mesh_data/hearth_physical_region.pvd") << subdomains
File("../data_files/mesh_data/hearth_facet_region.pvd") << boundaries